### PR TITLE
1735-automatic-deletion-of-connection-invitations

### PR DESCRIPTION
**Pull request by sahma19**
[Zenhub Issue](https://app.zenhub.com/workspaces/team-atlas-633199659e255a37cd1d144f/issues/gh/energinet-datahub/energy-origin-issues/1735)

### CODE REVIEW

File: domains/transfer-agreements/TransferAgreements.API/API.UnitTests/Services/ConnectionInvitationCleanupServiceTests.cs [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        var connectionInvitationRepository = Substitute.For<IConnectionInvitationRepository>();
+        var connectionInvitationCleanupService = new ConnectionInvitationCleanupService(logger, connectionInvitationRepository);
+
+        var cancellationTokenSource = new CancellationTokenSource();
+        var stoppingToken = cancellationTokenSource.Token;
+
+        var runTask = connectionInvitationCleanupService.Run(stoppingToken);
+
+        await Task.Delay(100, stoppingToken);
+        cancellationTokenSource.Cancel();
```
Comment (tnickelsen) [inline domains/transfer-agreements/TransferAgreements.API/API.UnitTests/Services/ConnectionInvitationCleanupServiceTests.cs:31]: why this?

File: domains/transfer-agreements/TransferAgreements.API/API.UnitTests/Services/ConnectionInvitationCleanupServiceTests.cs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+using System.Threading.Tasks;
+using API.Data;
+using API.Models;
+using API.Services.ConnectionInvitationCleanup;
+using FluentAssertions;
+using Microsoft.Extensions.Logging;
+using NSubstitute;
+using Xunit;
+
+namespace API.UnitTests.Services;
```
Comment (tnickelsen) [inline domains/transfer-agreements/TransferAgreements.API/API.UnitTests/Services/ConnectionInvitationCleanupServiceTests.cs:13]: Where are the integration tests?

File: domains/transfer-agreements/TransferAgreements.API/API/Data/IConnectionInvitationRepository.cs [hunk 3]
```diff
@@ -6,4 +7,6 @@ namespace API.Data;
 public interface IConnectionInvitationRepository
 {
     Task AddConnectionInvitation(ConnectionInvitation connectionInvitation);
+
+    Task<int> DeleteOldConnectionInvitations(TimeSpan olderThan);
```
Comment (tnickelsen) [inline domains/transfer-agreements/TransferAgreements.API/API/Data/IConnectionInvitationRepository.cs:11]: No need to return int

File: domains/transfer-agreements/TransferAgreements.API/API/Services/ConnectionInvitationCleanup/ConnectionInvitationCleanupService.cs [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+            try
+            {
+                await connectionInvitationRepository.DeleteOldConnectionInvitations(TimeSpan.FromDays(14));
+            }
+            catch (Exception e)
+            {
+                logger.LogWarning("Something went wrong with the InvitationCleanupService: {Exception}", e);
+            }
+
+            await SleepToNearestHour(stoppingToken);
```
Comment (tnickelsen) [inline domains/transfer-agreements/TransferAgreements.API/API/Services/ConnectionInvitationCleanup/ConnectionInvitationCleanupService.cs:38]: Maybe once every day instead?

**Review by tnickelsen — CHANGES_REQUESTED** [review]

File: domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/ConnectionInvitationsControllerTests.cs [hunk 5]
```diff
@@ -46,4 +42,77 @@ public async Task CreateConnectionInvitation_ShouldReturnUnauthorized_WhenUnauth
...
+    {
+        var invitationId = Guid.NewGuid();
+        var invitation = new ConnectionInvitation
+        {
+            Id = invitationId,
+            SenderCompanyId = Guid.NewGuid(),
+            SenderCompanyTin = "12345678"
+        };
+
+        await factory.SeedConnectionInvitations(new List<ConnectionInvitation> { invitation });
```
Comment (tnickelsen) [inline domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/ConnectionInvitationsControllerTests.cs:57]: No need to seed - you can call POST first then GET after

File: domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/ConnectionInvitationsControllerTests.cs [hunk 6]
```diff
@@ -46,4 +42,77 @@ public async Task CreateConnectionInvitation_ShouldReturnUnauthorized_WhenUnauth
...
+        var response = await client.GetAsync($"api/connection-invitations/{nonExistentInvitationId}");
+
+        response.StatusCode.Should().Be(HttpStatusCode.NotFound);
+
+        var responseBody = await response.Content.ReadAsStringAsync();
+        responseBody.Should().Be("Connection-invitation expired or deleted");
+    }
+
+    [Fact]
+    public async Task GetConnectionInvitation_ShouldReturnConflict_WhenConflictExists()
```
Comment (tnickelsen) [inline domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/ConnectionInvitationsControllerTests.cs:86]: _WhenConnectionExists

File: domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Services/ConnectionInvitationCleanupServiceTests.cs [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+            CreatedAt = DateTimeOffset.UtcNow.AddDays(-14)
+        };
+
+        dbContext.ConnectionInvitations.Add(newInvitation);
+        dbContext.ConnectionInvitations.Add(oldInvitation);
+        await dbContext.SaveChangesAsync();
+
+        var cancellationTokenSource = new CancellationTokenSource();
+
+        var runTask = cleanupService.Run(cancellationTokenSource.Token);
```
Comment (tnickelsen) [inline domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Services/ConnectionInvitationCleanupServiceTests.cs:60]: Shouldn't be necessary. The factory should run it

File: domains/transfer-agreements/TransferAgreements.API/API/Controllers/ConnectionInvitationsController.cs [hunk 8]
```diff
@@ -33,6 +39,37 @@ public async Task<ActionResult> CreateConnectionInvitation()
 
         await connectionInvitationRepository.AddConnectionInvitation(newInvitation);
 
-        return Ok(new { connectionInvitationId = newInvitation.Id });
+        return CreatedAtAction(nameof(GetConnectionInvitation), new { id = newInvitation.Id }, new { connectionInvitationId = newInvitation });
```
Comment (tnickelsen) [inline domains/transfer-agreements/TransferAgreements.API/API/Controllers/ConnectionInvitationsController.cs:42]: Why both new { id = newInvitation.Id } and new { connectionInvitationId = newInvitation }?

File: domains/transfer-agreements/TransferAgreements.API/API/Data/ConnectionInvitationRepository.cs [hunk 9]
```diff
@@ -14,4 +17,22 @@ public async Task AddConnectionInvitation(ConnectionInvitation connectionInvitat
...
+            .Where(i => i.CreatedAt < olderThan);
+
+        context.ConnectionInvitations.RemoveRange(oldConnectionInvitations);
+
+        await context.SaveChangesAsync();
+    }
+
+    public async Task<ConnectionInvitation?> FindConnectionInvitation(Guid id)
+    {
+        var connectionInvitation = await context.ConnectionInvitations
```
Comment (tnickelsen) [inline domains/transfer-agreements/TransferAgreements.API/API/Data/ConnectionInvitationRepository.cs:33]: Leave out expired invitations as well. Call it GetNonExpiredInvitation

**Review by tnickelsen — CHANGES_REQUESTED** [review]

File: domains/transfer/Transfer.API/API.IntegrationTests/Controllers/ConnectionInvitationsControllerTests.cs [hunk 10]
```diff
@@ -46,4 +43,78 @@ public async Task CreateConnectionInvitation_ShouldReturnUnauthorized_WhenUnauth
 
         result.StatusCode.Should().Be(HttpStatusCode.Unauthorized);
     }
+
+    [Fact]
+    public async Task GetConnectionInvitation_ShouldReturnOK_WhenInvitationExists()
```
Comment (tnickelsen) [inline domains/transfer/Transfer.API/API.IntegrationTests/Controllers/ConnectionInvitationsControllerTests.cs:48]: Hvad blev der af den test, der oprettede via POST og derefter brugte GET til at verificere oprettelse?

Comment (sahma19) [reply domains/transfer/Transfer.API/API.IntegrationTests/Controllers/ConnectionInvitationsControllerTests.cs:50]: Fixed in https://github.com/Energinet-DataHub/energy-origin/pull/415/commits/f040ef535680294e0a47ae5624b620361a4cebcf

File: domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/ConnectionInvitationsControllerTests.cs (hunk 5 above)
Comment (sahma19) [reply domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/ConnectionInvitationsControllerTests.cs]: fixed in https://github.com/Energinet-DataHub/energy-origin/pull/415/commits/f040ef535680294e0a47ae5624b620361a4cebcf

File: domains/transfer/Transfer.API/API.IntegrationTests/NpgsqlConnectionExtension.cs [hunk 11]
```diff
@@ -0,0 +1,31 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.EntityFrameworkCore;
+
+namespace API.IntegrationTests;
+
+public static class NpgsqlConnectionExtension
```
[the remark this answers is not in the document]
Comment (sahma19) [reply domains/transfer/Transfer.API/API.IntegrationTests/NpgsqlConnectionExtension.cs]: Fixed in https://github.com/Energinet-DataHub/energy-origin/pull/415/commits/53b28509b6a6d9b2ceab4919c6432e268f3caa6e

**Review by tnickelsen — APPROVED** [review]

**Pull request merged**